### PR TITLE
Fix multi-window interaction with the single process work

### LINKED ISSUE #968: Regression: Opening a second instance of the App doesn't create a new Window
Regression introduced with #877

When opening a second instance of the App the current mainwindow is brought into focus instead of creating a second main window.

Reproduced this on Windows 11.
Edit: erri confirmed same issue on Linux as well.

Use manually started instances of the App to reproduce this, starting the App with debugger makes it not reachable by normal instances of the app, meaning the bug would not reproduce.

### PR DESCRIPTION
**Pull request by halgari**
Fixes #968 and #965. 

This fixes the above bugs by reworking how we open multiple windows. Namely:

* Whenever a window opens, create a main process director token to keep the single process director alive
* Startup Avalonia in a mode that only shuts down when a cancellation token exits
* Link that cancellation token to the single process director cancellation token
* Only run the Avalonia startup and the first window on the creation thread, run the rest via the UI dispatcher (this fixes the #968 regression)

### CODE REVIEW

File: src/NexusMods.App.UI/Windows/MainWindowViewModel.cs [hunk 1]
```diff
@@ -64,11 +64,6 @@ public class MainWindowViewModel : AViewModel<IMainWindowViewModel>, IMainWindow
 
         this.WhenActivated(d =>
         {
-            // When the user closes the window, we should persist all download state such that it shows
-            // accurate values after a reboot.
-            // If we ever plan to remove and re-add main window (unlikely), this might need changing to not dispose but only save.
-            downloadService.DisposeWith(d);
```
Comment (halgari) [inline src/NexusMods.App.UI/Windows/MainWindowViewModel.cs:70]: This shouldn't be needed, the service is disposed by DI on shutdown

File: src/NexusMods.App/MainThreadData.cs [hunk 2]
```diff
@@ -9,7 +9,22 @@ public static class MainThreadData
...
+    /// </summary>
+    public static CancellationToken GlobalShutdownToken => _cancellationTokenSource.Token;
+    
+    /// <summary>
+    /// Shuts down the application.
+    /// </summary>
+    public static void Shutdown()
+    {
+        _cancellationTokenSource.Cancel();
+    }
```
Comment (erri120) [inline src/NexusMods.App/MainThreadData.cs:26]: `Cancel` will throw all aggregated exceptions, those should be handled somewhere.

`OperationCanceledException` instances can be ignored.

File: src/NexusMods.App/Startup.cs [hunk 3]
```diff
@@ -17,9 +21,27 @@ namespace NexusMods.App;
...
+
+#pragma warning disable CS0028 // Disables warning about not being a valid entry point
+    
+    /// <summary>
+    /// Main entry point for the application, the application loop will only exit when the token is cancelled.
+    /// </summary>
+    /// <param name="provider"></param>
+    /// <param name="args"></param>
+    /// <param name="token"></param>
     public static void Main(IServiceProvider provider, string[] args)
```
Comment (erri120) [inline src/NexusMods.App/Startup.cs:36]: Empty `param` XML docs should be removed, they just clutter up the code.

File: src/NexusMods.App/Startup.cs [hunk 4]
```diff
@@ -41,14 +66,45 @@ public static void Main(IServiceProvider provider, string[] args)
...
 
+    private static void AppMain(Application app, string[] args)
+    {
+        _logger.LogInformation("Starting UI application");
+        ShowMainWindow();
+
+        // Start the app, linking to the global shutdown token
+        _logger.LogInformation("Starting application loop.");
+        app.Run(MainThreadData.GlobalShutdownToken);
+        _logger.LogInformation("Application loop ended.");
```
Comment (erri120) [inline src/NexusMods.App/Startup.cs:77]: Log event messages should be fragments, not sentences. Avoid a trailing period/full stop.

File: src/NexusMods.App/Program.cs [hunk 5]
```diff
@@ -54,13 +54,24 @@ public static int Main(string[] args)
...
 
 
         // The UI *must* be started on the main thread, according to the Avalonia docs, although it
         // seems to work fine on some platforms (this behavior is not guaranteed). So when we need to open a new
         // window, the handler will enqueue an action to be run on the main thread.
-        while (!managerTask.IsCompleted)
+        while (!MainThreadData.GlobalShutdownToken.IsCancellationRequested)
         {
             if (MainThreadData.MainThreadActions.TryDequeue(out var action))
             {
```
Comment (erri120) [inline src/NexusMods.App/Program.cs:77]: The `action()` call inside the loop should probably be surrounded by a `try/catch`.

**Review by erri120 — CHANGES_REQUESTED** [review]
See comments.

The app now doesn't open anymore, it like the following exists immediately:

```csharp
_logger.LogInformation("Starting application loop.");
app.Run(MainThreadData.GlobalShutdownToken);
_logger.LogInformation("Application loop ended.");
```

```log
00:00:01.793 [INFO] Starting UI application
00:00:02.104 [INFO] Starting application loop.
00:00:02.105 [INFO] Application loop ended.
00:00:02.105 [DEBUG] Startup handler returned 0
```

Edit: `MainThreadData.GlobalShutdownToken` appears to already be canceled.

**Comment by halgari** [comment]
Is this in debug mode? Works fine in debug without a debugger attached on my machine (`dotnet run`). Looks like with a debugger attached it no longer works, I'll look into that

**Comment by erri120** [comment]
> Is this in debug mode? Works fine in debug without a debugger attached on my machine (`dotnet run`). Looks like with a debugger attached it no longer works, I'll look into that

https://github.com/Nexus-Mods/NexusMods.App/blob/9232312b074836fa3aeedb4199b55f79c7072d50/src/NexusMods.App/StartupHandler.cs#L50-L63

`tcs.SetResult(0)` will "finish" `StartUiWindowAsync` which finishes `await startup.Start(args, _isDebug);` in `Program.cs`.

File: src/NexusMods.App/Program.cs (hunk 5 above)
Comment (halgari) [reply src/NexusMods.App/Program.cs:82]: handled

**Comment by halgari** [comment]
This gets us up back and running. I don't like how there's two paths for the code now, where some things only happen under Debug mode, but I need to think how to better manage this in a way where the code is a bit cleaner. Perhaps a lot of this logic can be moved into the `SingleProcess` project

**Review by erri120 — CHANGES_REQUESTED** [review]
Multiple windows now work again, however, you're unable to close any windows. If you close a window, an exception is thrown due to https://github.com/AvaloniaUI/Avalonia/issues/14437 and the window stays frozen on screen. The only way to solve this is by manually killing the main process of the app.

**Comment by halgari** [comment]
What process are you going through to get this to happen? I just opened and closed a bunch of windows with this branch and didn't see this exception?

**Comment by erri120** [comment]
> What process are you going through to get this to happen? I just opened and closed a bunch of windows with this branch and didn't see this exception?

The exception should always appear in the log, since the underlying issue https://github.com/AvaloniaUI/Avalonia/issues/14437 hasn't been fixed yet.

However, the last two commits of this PR introduced another issue: the window doesn't close properly, and the main process is still running. I had to manually kill the main process before the windows disappeared. This happens when running with a debugger and without.

**Comment by halgari** [comment]
I can't get this error at all, it all cleanly shuts down for me, I'll switch over to my ubuntu machine and see if I can get it to happen there.

**Comment by halgari** [comment]
Ah I haven't managed a game yet, let me try that

**Comment by Al12rs** [comment]
@halgari The app crash on close issue was fixed in:
- #1068 

That should unblock work on this PR again.
The changes from main need to be merged in.

**Review by erri120 — CHANGES_REQUESTED** [review]
Running the app without an attached debugger via `dotnet run`, the processes remain running after closing the window.

When running the app with an attached debugger and running two instances, you get two processes, not a single process.

**Comment by Sewer56** [comment]
When Running on Linux, this PR makes it so you can't run the App without the debugger attached.

First process spawns 2nd process, then it dies, as far as I can tell.

**Pull request merged**